book-2-platform-api/chapters/API_MODELS_CONTROLLERS.md

```diff
@@ -96,70 +158,120 @@ namespace CoffeeShop.Controllers
+                        };
+
+                        coffees.Add(coffee);
+                    }
+                    reader.Close();
 
+                    return Ok(coffees);
+                }
+            }
         }
 
         [HttpGet("{id}", Name = "GetCoffee")]
         public async Task<IActionResult> Get([FromRoute] int id)
         {
-            using (IDbConnection conn = Connection)
+            using (SqlConnection conn = Connection)
             {
-                string sql = $"SELECT * FROM Coffee WHERE Id = {id}";
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"
+                        SELECT
+                            Id, Title, BeanType
+                        FROM Coffee
+                        WHERE Id = @id";
+                    cmd.Parameters.Add(new SqlParameter("@id", id));
+                    SqlDataReader reader = cmd.ExecuteReader();
 
-                var singleCoffee = (await conn.QueryAsync<Coffee>(sql)).Single();
-                return Ok(singleCoffee);
+                    Coffee coffee = null;
+
+                    if (reader.Read())
+                    {
+                        coffee = new coffee
+                        {
+                            Id = reader.GetInt32(reader.GetOrdinal("Id")),
+                            Title = reader.GetString(reader.GetOrdinal("Title")),
+                            BeanType = reader.GetString(reader.GetOrdinal("BeanType"))
+                        };
+                    }
+                    reader.Close();
+
+                    return Ok(coffee);
+                }
             }
         }
 
         [HttpPost]
         public async Task<IActionResult> Post([FromBody] Coffee coffee)
         {
-            string sql = $@"INSERT INTO Coffee
-            (Title, BeanType)
-            VALUES
-            ('{coffee.Title}', '{coffee.BeanType}');
-            select MAX(Id) from Coffee";
-
-            using (IDbConnection conn = Connection)
+            using (SqlConnection conn = Connection)
             {
-                var newId = (await conn.QueryAsync<int>(sql)).Single();
-                coffee.Id = newId;
-                return CreatedAtRoute("GetCoffee", new { id = newId }, coffee);
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"INSERT INTO Coffee (Title, BeanType)
+                                        OUTPUT INSERTED.Id
+                                        VALUES (@title, @beanType)";
+                    cmd.Parameters.Add(new SqlParameter("@title", coffee.Title));
+                    cmd.Parameters.Add(new SqlParameter("@beanType", coffee.BeanType));
+                    cmd.Parameters.Add(new SqlParameter("@id", id));
+
+                    cmd.ExecuteNonQuery();
```

Comment:
This `OUTPUT INSERTED.Id` thing is cool.

This SO post suggests that we should only use `cmd.ExecuteScalar()` alone and not have `cmd.ExecuteNonQuery()`
https://stackoverflow.com/questions/10999396/how-do-i-use-an-insert-statements-output-clause-to-get-the-identity-value

Reply:
True. Gone.